pkg/controlplane/reconcilers/peer_endpoint_lease.go

```diff
@@ -0,0 +1,426 @@
+	"net"
+	"path"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"k8s.io/klog/v2"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/features"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/apiserver/pkg/storage/storagebackend"
+	storagefactory "k8s.io/apiserver/pkg/storage/storagebackend/factory"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	endpointsv1 "k8s.io/kubernetes/pkg/api/v1/endpoints"
+)
+
+const (
+	APIServerIdentityLabel = "apiserverIdentity"
+)
+
+// PeerLeases is an interface which assists in managing the set of active servers
+type PeerLeases interface {
+	// ListLeases retrieves a list of the current server IPs
+	ListLeases() ([]string, error)
+
+	// UpdateLease adds or refreshes a servers lease
+	UpdateLease(ip string, peerBindAddress string, serverId string, endpointPorts []corev1.EndpointPort) error
+
+	// GetLease retrieves the IP of a specific service
+	GetLease(serverId string) (string, error)
+
+	// RemoveLease removes a server's lease
+	RemoveLease(ip string, endpointPorts []corev1.EndpointPort) error
+
+	// Destroy cleans up everything on shutdown.
+	Destroy()
+}
+
+type peerEndpointLeases struct {
+	storage   storage.Interface
+	destroyFn func()
+	baseKey   string
+	leaseTime time.Duration
+}
+
+// ListLeases retrieves a list of the current server IPs from storage
+func (s *peerEndpointLeases) ListLeases() ([]string, error) {
+	storageOpts := storage.ListOptions{
+		ResourceVersion:      "0",
+		ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan,
+		Predicate:            storage.Everything,
+		Recursive:            true,
+	}
+	ipInfoList, err := s.GetIpInfoList(storageOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	ipList := make([]string, 0, len(ipInfoList.Items))
+	for _, ip := range ipInfoList.Items {
+		if len(ip.Subsets) > 0 && len(ip.Subsets[0].Addresses) > 0 && len(ip.Subsets[0].Addresses[0].IP) > 0 {
+			ipList = append(ipList, ip.Subsets[0].Addresses[0].IP)
+		}
+	}
+
+	klog.V(6).Infof("Current server IPs listed in storage are %v", ipList)
+
+	return ipList, nil
+}
+
+// UpdateLease resets the TTL on a server IP in storage
+// UpdateLease will create a new key if it doesn't exist.
+// It will also do the following if UnknownVersionInteroperabilityProxy feature is enabled
+// 1. store the apiserverId as a label if
+// 2. store the values passed to --peer-bind-ip and --peer-bind-port flags to kube-apiserver as an annotation
+func (s *peerEndpointLeases) UpdateLease(ip string, peerBindAddress string, serverId string, endpointPorts []corev1.EndpointPort) error {
+	key := path.Join(s.baseKey, ip)
+	// TODO: how to use endpointPorts to save in the key
+	if len(endpointPorts) > 0 {
+		key = path.Join(key, fmt.Sprint(endpointPorts[0].Port))
+	}
```

Comment:
do we have two apiservers with the same ip?

Reply:
the integration test does, yes. It creates 2 test servers on the same ip but different ports. 
I dont have a clear idea on how to make the integration test work without doing this (modifying lease's base key to include ports), but it does seem like this is not a real world scenario: having 2 apiservers with same ip.

Reply:
resolving, since now we are using apiserver IDs for lease keys.